src/vm/methodtable.cpp

```diff
@@ -3092,8 +3103,25 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
                 currentEightByte++;
                 _ASSERTE(currentEightByte <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
 
-                currentEightByteOffset = offset + fieldSize;
-                accumulatedSizeForEightByte = 0;
+                currentEightByteOffset += SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
+                accumulatedSizeForEightByte -= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
+
+                // If a field is large enough to span multiple eightbytes, then set the eightbyte classification to the field's classification.
+                if (accumulatedSizeForEightByte > 0)
+                {
+                    helperPtr->eightByteClassifications[currentEightByte] = fieldClassificationType;
+                    helperPtr->eightByteSizes[currentEightByte] = accumulatedSizeForEightByte;
+                    helperPtr->eightByteOffsets[currentEightByte] = currentEightByteOffset;
+
+                    _ASSERTE(accumulatedSizeForEightByte <= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES); // This code path only supports fields up to 2 eightbytes (the max number as of writing)
+                    if(accumulatedSizeForEightByte == SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES)
```

Comment:
Hmm, it got pretty complex after all the updates. I was thinking that we would update the code at line 2127 and below to handle this case (that's why I was suggesting to change the `<` to `<=`). Basically, that code would handle the "overflow" from the main loop. That means three cases:
1. The struct is smaller than one eightbyte and so it ends in the middle of it
2. The struct is larger than one eightbyte but smaller than two eightbytes, so it ends in the middle of the second eightbyte
3. The struct size is exactly two eightbytes and its last field has crossed eightbyte boundary.

So the code would become something like:
```C++
    if (accumulatedSizeForEightByte >= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES)
    {
        // Save data for this eightbyte.
        helperPtr->eightByteSizes[currentEightByte] = SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
        helperPtr->eightByteOffsets[currentEightByte] = currentEightByteOffset;
    
        // Set up for next eightbyte.
        currentEightByte++;
        _ASSERTE(currentEightByte <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
    
        currentEightByteOffset += SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
        accumulatedSizeForEightByte -= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
    
        // If a field is large enough to span multiple eightbytes, then set the eightbyte classification to the field's classification.
        if (accumulatedSizeForEightByte > 0)
        {
            helperPtr->eightByteClassifications[currentEightByte] = fieldClassificationType;
        }
    }
}

_ASSERTE(accumulatedSizeForEightByte < SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES);
if (accumulatedSizeForEightByte > 0 && accumulatedSizeForEightByte <= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES)
{
    helperPtr->eightByteSizes[currentEightByte] = accumulatedSizeForEightByte;
    helperPtr->eightByteOffsets[currentEightByte] = currentEightByteOffset;
    currentEightByte++;
}
```

Alternatively, we can get back to your original solution with the while loop here, updated a bit:
```C++
    while (accumulatedSizeForEightByte >= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES)
    {
        // Save data for this eightbyte.
        helperPtr->eightByteSizes[currentEightByte] = SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
        helperPtr->eightByteOffsets[currentEightByte] = currentEightByteOffset;
        
        // Set up for next eightbyte.
        currentEightByte++;
        _ASSERTE(currentEightByte <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
        
        currentEightByteOffset += SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
        accumulatedSizeForEightByte -= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES;
        
        // If a field is large enough to span multiple eightbytes, then set the eightbyte classification to the field's classification.
        if (accumulatedSizeForEightByte > 0)
        {
            helperPtr->eightByteClassifications[currentEightByte] = fieldClassificationType;
        }
    }
}
```

Reply:
I think I'm going to go for the loop. I personally think it looks cleaner.